### ISSUE TITLE
Creating a new user with password

### ISSUE THREAD

**Issue by CrestApps** [OrchardCMS/OrchardCore#11080]
### Is your feature request related to a problem? Please describe.
By default apps use basic-authentication to authenticate users into their site including Microsoft Identity based apps. No user is able to authenticate/login without providing "username" and "password". Alternatively, an admin could enable some sort of external authentication providers like Azure AD or any other providers to avoid the need to manage the authentication by the app directly. These external providers are optional and not enabled by default in OC so creating a user with no password by default adds no value since no the user can't login without one. 

Currently, when adding a new user using the admin UI, the admin has to create a user and then go to a second view to set/reset their password. This is a very awkward workflow specially if you want to add many users. Since a username/password are required by default for a user to login, I think the admin should be required to provide a password rather than creating a useless user with no password. Note, the point here is that password is required to login by default without having to enable optional additional authentication providers.

### Describe the solution you'd like
I think we should add a security settings something like "IsPasswordRequired" in addition to [PasswordOptions ](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.identity.passwordoptions?view=aspnetcore-6.0) setting for the user to control. If no external authentication provider is provided, the Password should be required "otherwise the user can't login without password". Now in the [CreatePost](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Users/Controllers/AdminController.cs#L319-L345) action we can check if a password is required then validate password and create a user with password. Otherwise, we create the user with no password since password is externally managed. I would think that the controller code will then look something like this

```
public async Task<IActionResult> CreatePost(string password)
{
    var user = new User();

    if (!await _authorizationService.AuthorizeAsync(User, Permissions.ViewUsers, user))
    {
        return Forbid();
    }

    var shape = await _userDisplayManager.UpdateEditorAsync(user, updater: _updateModelAccessor.ModelUpdater, isNew: true);

    if (!ModelState.IsValid)
    {
        return View(shape);
    }

    if (setting.IsPasswordRequiredForNewUsers)
    {
        // I am guessing the password will be passed to Identity for validation before the user is created
        // the ModelState would get updated if the password does not meet the standard defined using PasswordOptions in identity
        await _userService.CreateUserAsync(user, password, ModelState.AddModelError);
    } 
    else 
    {
        await _userService.CreateUserAsync(user, null, ModelState.AddModelError);
    }

    if (!ModelState.IsValid)
    {
        return View(shape);
    }

    await _notifier.SuccessAsync(H["User created successfully."]);

    return RedirectToAction(nameof(Index));
}
```

At this point the process becomes more robust. If a user want to generate a random password, then the password shape will generate a random password or any other password requirement is required. And if the user want to require the user to set new password after the first login, then he/she can create a display driver `DisplayDriver<User>` that would allow them to inject the option and require the user to change their password after they login,

I would like to add "Password" field on the Create User screen and create the new user with a password instead of null.

### Describe alternatives you've considered
I think the only way I can do this is to create a new module that depends on `OrchardCore.User` and override the `/admin/user/create` route with my own code. Now sure if this wold even work

**Comment by CrestApps** [comment]
> @CrestApps as admin you can always edit password once user is created.
> 
> IMO if you need to update the password when user is created ( whether it's auto generated for post back) you can achieve using workflow - UserCreatedEvent

I think the app should have a default behavior without the need to adding anything else like workflow. Since the default authentication provider is basic-authentication, I think password should be part of the default process with the ability to change it if someone want to extend the default behavior. Adding a setting will make this process more robust

**Comment by Skrypt** [comment]
A custom handler for your own requirements might be what you are looking for.

**Comment by ns8482e** [comment]
Password is not required for all users in all scenarios ( except admin while setup the site ) so including password field on create/edit screen doesn't really useful

Also,  While editing user - you really don't want to edit the password on same form

**Comment by Skrypt** [comment]
One thing that bothers me is that the password is generated still for these Users. And there is currently no way to inform them about that password by email. You need to create a Workflow, but the workflow cannot decrypt the generated password. So, you need to manually set a password or let them reset their password on the "login" page.

**Comment by CrestApps** [comment]
@ns8482e the password field should only show up on the create user view not the edit users. And you are right password is not required for all scenarios but is required for the default scenario (basic-authentication).

@Skrypt I do not see event being fired before the User is created and after the user is created to be able to validation password before and then set the password after the user is created. Validating the password should happen before the user is created and setting the password would be done after in events is the way to go.

**Comment by Skrypt** [comment]
ASP.NET Identity Framework allows to create users without a Password. And I think you can make it work that users can log in without a password if they want to. So, for OC here, the idea is to be able to cover a lot of possible scenarios. Though, when we create a user without a password; the `PasswordHash` in the database is set to NULL.

https://stackoverflow.com/questions/33813376/creating-users-with-no-password-using-asp-net-identity

Though, here, what you want is to generate a Password for them when you create a User for them from the Admin UI. This means that right now you need to override this OrchardCore.Users Admin Controller to do what you want. Creating a Workflow would execute whether a user is created from the frontend or from the backend by an admin. So, there is no distinction unless in your handler you verify that the requests come from the backend only.

**Comment by Skrypt** [comment]
I've read your proposal a second time and I think that you need to understand that the reason the Password is not mandatory is because of the fact that ASP.NET Identity doesn't require one. Though, I agree that it is misleading that we can create a User without a password and that by default we can't login without a password. I think this is the issue.

**Comment by CrestApps** [comment]
@Skrypt yes. I am aware that Identity allows you to create a user without password because in some cases when other authentication providers are enabled, the password is managed externally thus a password is not needed. But OC is shipped with a basic authentication (username and password) and for that reason we should make this options configurable "based on the user needed" but mandate password by default "during the creating of the user".

**Comment by Skrypt** [comment]
Though this is not based on the authentication provider used. It's just that creating users without passwords was more universal by the time it got implemented. I agree that there is something that needs to be done here to make this more user-friendly. Though, as you can see, at least there are workarounds. If you want to contribute to a PR then start one. Though we need to agree on the solution first.

**Comment by CrestApps** [comment]
@Skrypt once we agree on the solution, I'll attempt a new PR.

**Comment by ns8482e** [comment]
@Skrypt / @CrestApps  when used external identity - Password is not required and creating local password is optional. Also admin can disable local login from User login settings.

Adding one more flag (`IsPasswordRequiredForNewUsers` and admin has to maintain it)  just for showing password field on create user screen is counter intuitive. 

solution may be just create in two steps - i.e. add "Save and Create Password" button, ( shape )  so it will redirect to Create Password screen after creating  user.

**Comment by sebastienros** [comment]
Since there is an "Edit Password" this should not be an issue to enter it on the user creation screen. To "not provide a password" it could be based on not entering data in the fields, or provide a check box "Edit password". This would prevent from having to open two windows.

**Comment by Skrypt** [comment]
@CrestApps you have your answer 😉

**Comment by CrestApps** [comment]
> @CrestApps you have your answer 😉

I don't think I follow what is the outcome here. 

If I understand the comment here, I would create a `PasswordPart` that would look something like this

```
public class PasswordPart : ContentPart
{
    public bool HasPassword { get; set; }
    public string Password { get; set; }
    public string ConfirmPassword { get; set; }
}
```

Then using either update [UserDisplayDriver](https://github.com/OrchardCMS/OrchardCore/blob/c8b2954e038aaa54963b60bc6cfe193cfea54799/src/OrchardCore.Modules/OrchardCore.Users/Drivers/UserDisplayDriver.cs) or create a new driver that inherit from `DisplayDriver<User>` to attach the part. But if that is the case, how would I pass the password to the `IUserService` during [creating the user](https://github.com/OrchardCMS/OrchardCore/blob/c8b2954e038aaa54963b60bc6cfe193cfea54799/src/OrchardCore.Modules/OrchardCore.Users/Controllers/AdminController.cs#L335) or editing the users?

**Comment by sebastienros** [comment]
Not a new part, just change the user creation view to have a password field as described.
